safenode/src/node/api.rs

```diff
@@ -113,118 +112,123 @@ impl Node {
-        &mut self,
-        signed_spend: Box<SignedSpend>,
-        source_tx: Box<DbcTransaction>,
-        response_channel: ResponseChannel<Response>,
-    ) -> Result<()> {
-        // Ensure that the provided src tx is the same as the one we have the hash of in the signed spend.
-        let provided_src_tx_hash = source_tx.hash();
-        let signed_src_tx_hash = signed_spend.src_tx_hash();
-
-        if provided_src_tx_hash != signed_src_tx_hash {
-            let error = ProtocolError::SignedSrcTxHashDoesNotMatchProvidedSrcTxHash {
-                signed_src_tx_hash,
-                provided_src_tx_hash,
-            };
-            self.send_response(
-                Response::Cmd(CmdResponse::Spend(Err(error))),
-                response_channel,
-            )
-            .await;
-            return Ok(());
+    async fn handle_query(&mut self, query: Query) -> QueryResponse {
+        match query {
+            Query::Register(query) => self.registers.read(&query, User::Anyone).await,
+            Query::GetChunk(address) => {
+                let resp = self.chunks.get(&address).await;
+                QueryResponse::GetChunk(resp)
+            }
+            Query::Spend(query) => {
+                match query {
+                    SpendQuery::GetFees { dbc_id, priority } => {
+                        // The client is asking for the fee to spend a specific dbc, and including the id of that dbc.
+                        // The required fee content is encrypted to that dbc id, and so only the holder of the dbc secret
+                        // key can unlock the contents.
+                        let required_fee = self.transfers.get_required_fee(dbc_id, priority);
+                        QueryResponse::GetFees(Ok(required_fee))
+                    }
+                    SpendQuery::GetDbcSpend(address) => {
+                        let res = self
+                            .transfers
+                            .get(address)
+                            .await
+                            .map_err(ProtocolError::Transfers);
+                        QueryResponse::GetDbcSpend(res)
+                    }
+                }
+            }
         }
+    }
 
-        // First we need to validate the parents of the spend.
-        // This also ensures that all parent's dst tx's are the same as the src tx of this spend.
-        match self
-            .validate_spend_parents(signed_spend.as_ref(), source_tx.as_ref())
-            .await
-        {
-            Ok(()) => (),
-            Err(super::Error::Protocol(error)) => {
-                // We drop spend attempts with invalid parents,
-                // and return an error to the client.
-                self.send_response(
-                    Response::Cmd(CmdResponse::Spend(Err(error))),
-                    response_channel,
-                )
-                .await;
-                return Ok(());
+    async fn handle_cmd(&mut self, cmd: Cmd) -> CmdResponse {
```

Comment:
Probably not something to be addressed in this PR, but it could be worth thinking about unit testing the command and event handlers. The storage and so on is something it would be possible to mock.